components/merkledb/examples/migration/mod.rs

```diff
@@ -181,112 +193,89 @@ mod v2 {
-
-    let fork = db.fork();
-    let new_data = Migration::new("test", &fork);
-    let old_data = Prefixed::new("test", fork.readonly());
-    {
-        let old_schema = v1::Schema::new(old_data.clone());
-        println!("Before migration:");
-        old_schema.print_wallets();
-    }
-    create_migration(new_data, old_data);
-    db.merge(fork.into_patch()).unwrap();
-
-    // For now, the old data is still present in the storage.
-    let snapshot = db.snapshot();
-    let old_schema = v1::Schema::new(Prefixed::new("test", &snapshot));
+/// Checks that we have old and new data in the storage after migration.
+fn check_data_before_flush(snapshot: &dyn Snapshot) {
+    let old_schema = v1::Schema::new(Prefixed::new("test", snapshot));
     assert_eq!(old_schema.ticker.get().unwrap(), "XNM");
     // The new data is present, too, in the unmerged form.
-    let new_schema = v2::Schema::new(Migration::new("test", &snapshot));
+    let new_schema = v2::Schema::new(Migration::new("test", snapshot));
     assert_eq!(new_schema.config.get().unwrap().ticker, "XNM");
 
-    let system_schema = SystemSchema::new(&snapshot);
+    let system_schema = SystemSchema::new(snapshot);
     let state = system_schema.state_aggregator();
     assert_eq!(state.keys().collect::<Vec<_>>(), vec!["unrelated.list"]);
-    let migration_view = Migration::new("test", &snapshot);
+    let migration_view = Migration::new("test", snapshot);
     let state = migration_view.state_aggregator();
     assert_eq!(
         state.keys().collect::<Vec<_>>(),
         vec!["test.config", "test.wallets"]
     );
     let new_state_hash = state.object_hash();
     assert_eq!(new_state_hash, migration_view.state_hash());
+}
 
-    let mut fork = db.fork();
-    flush_migration(&mut fork, "test");
-    let patch = fork.into_patch();
-
-    // Now, the new indexes have replaced the old ones.
-    let new_schema = v2::Schema::new(Prefixed::new("test", &patch));
+/// Checks that old data was replaced by new data in the storage.
+fn check_data_after_flush(view: impl RawAccess + AsReadonly + Copy) {
+    let new_schema = v2::Schema::new(Prefixed::new("test", view));
     assert_eq!(new_schema.config.get().unwrap().divisibility, 8);
-    assert!(!patch.get_entry::<_, u8>("test.divisibility").exists());
+    assert!(!view.get_entry::<_, u8>("test.divisibility").exists());
 
     // The indexes are now aggregated in the default namespace.
-    let system_schema = SystemSchema::new(&patch);
+    let system_schema = SystemSchema::new(view);
     let state = system_schema.state_aggregator();
     assert_eq!(
         state.keys().collect::<Vec<_>>(),
         vec!["test.config", "test.wallets", "unrelated.list"]
     );
+}
+
+/// Performs common migration logic.
+pub fn perform_migration<F>(migrate: F)
+where
+    F: FnOnce(Arc<dyn Database>),
+{
+    // Creating a temporary DB and filling it with some data.
+    let db: Arc<dyn Database> = Arc::new(create_initial_data());
```

Comment:
Why you not just return `Arc<Database>` from the `create_initial_data` function?

Reply:
I've suggested to return `TemporaryDB`. The motivation is that it's possible to use `Arc<TemporaryDB>` here and coerce it to `Arc<dyn Database>` for the `migrate` call; `Arc<TemporaryDB>` is marginally faster since it doesn't involve dynamic dispatch. I suppose performance is not the main concern for examples, so I wouldn't protest doing the things as you suggest 🙃